rfcs/require-by-string.md

```diff
@@ -0,0 +1,287 @@
+
+Additionally, if an alias is bound to a relative path, the path will evaluated relative to the `.luaurc` file in which the alias was defined.
+
+#### Path resolution
+
+If we find files with the same name but different extensions, then we will attempt to require a file with the following extensions (in this order):
+1. `.luau`
+2. `.lua`
+3. All other file extensions are invalid.
+
+If the string resolves to a directory instead of a file, then we will attempt to require a file in that directory with the following name (in this order):
+1. `init.luau`
+2. `init.lua`
+
+#### DataModel as VFS
+
+In the Roblox engine, the DataModel will act as a virtual file system. At the root of this VFS is the DataModel itself. This will allow packages from Luau to be imported and exported freely without needing to consider the platform they are being used on.
+
+All paths used in the Roblox engine must refer to a location in the DataModel. They cannot be used to access files on disk.
+
+```lua
+-- MyModule location: game:GetService("ReplicatedStorage").MyModule
+
+-- Require by absolute path
+local MyModule = require("/ReplicatedStorage/MyModule")
+
+-- Require by relative path
+-- (From game:GetService("ReplicatedStorage").MyOtherModule)
+local MyModule = require("MyModule")
+```
+
+##### Considerations
+
+Within the Roblox engine, we will have to handle replication and waiting for modules to be loaded into the DataModel before requiring them. However, this is outside of the scope of this RFC and will be discussed internally.
+
+#### Platform compatibility
+
+For compatibility across platforms, we will automatically map `/` onto `\`.
+
+## Drawbacks
+
+### Backwards compatibility
+
+Luau libraries are already not compatible with existing Lua libraries. This is because Lua favors the `.` based require syntax instead and relies on the `LUA_PATH` environment variable to search for modules, whereas Luau currently supports a basic require-by-string syntax.
+
+- Libraries are fully compatible with the Roblox engine, as require-by-string is currently unimplemented.
+- Luau currently implements relative paths in relation to the current working directory. 
+- We propose changing this behavior, and breaking backwards compatibility on this front.
+- With the current implementation, requiring a library that itself contains relative-path require statements [can become a mess](https://github.com/Roblox/luau/issues/959) if the Luau VM is not launched from the "correct" working directory.
+- We propose the following change: relative paths passed to require statements will be evaluated in relation to the requiring file's location, not in relation to the current working directory.
+    - Caveat: if the relative-path require is called directly in the Luau CLI, this will remain relative to the current working directory, as there is no sense of a "requiring file" here.
```

Comment:
I'm a bit confused by this in relation to the other parts of the proposal.

Does this proposal only apply to Luau type-checking and not the generic runtime?
While Roblox can have it's own internal runtime implementation it feels strange to only provide a public spec on how things should be resolved without including an implementation example in the Luau CLI (do we suggest using lua_State userdata to track original "cwd" of a thread that calls 'require' or do something else?)

Reply:
I think we'll need to discuss how we want the "cwd" of a thread to be stored before adding these implementation details to the RFC.

We can take the approach of storing the path as userdata/lightuserdata in the lua_State. @BradSharp also had proposed piggybacking off the internal mechanics of debug.info(), which can be used to obtain the script name in which a function is defined.

Reply:
@vegorov-rbx I've added details about retrieving the requiring file's location to the RFC: [Implementing changes to relative paths](https://github.com/VarunSaini02/luau/blob/667a42d2168d70b923b800adc99d37794a5d9cc9/rfcs/require-by-string.md?plain=1#L227).

Feel free to take a look when you have a chance.

Reply:
Since you added those details, why is this caveat still a thing in the spec?

Reply:
I just pushed a change clarifying that this caveat is specifically about the case in which a `require` statement is executed in a REPL input prompt.